### PR TITLE
feat(dsv4): integrate standalone decode MoE front

### PR DESCRIPTION
**Pull request by zsnoob**
## Summary

- Integrate the standalone four-kernel MoE front into the DSV4 Decode MegaMoE-SE path.
- Fuse mHC collapse, FP32 SSQ/RMSNorm, router projection, TopK, FP8 quantization, and publication into MegaMoE-SE buffers.
- Retain CUDA Graph plans only for graph-stable input addresses; eager plans remain temporary.
- Gate activation on the extension geometry, ABI, source identity, dependency identity, and PyTorch-reported CUDA capability.
- Preserve the ordinary MoE path for unsupported layouts, M > 256, non-MegaMoE-SE strategies, and non-int32 HashMoE input IDs.
- Keep the feature independent of CSA and HCA under DSV4_MEGA_MOE_FRONT.

## Scope

This PR is Decode-only. Prefill construction and execution remain unchanged. In P/D validation, Prefill is used only as the minimum request and KV-handoff entry and is not an optimization or measurement target.

The change is limited to the DSV4 MoE front integration and MegaMoE-SE publication boundary. It does not modify CSA, HCA, attention kernels, or token-level scheduling.

## Dependency contract

- CUDA extension: 70819b58db7350ae711dbe62a1a7b8f00b578765
- DeepGEMM: 559d79fb6994a58b8a15b4b93bf13ccc16edf247
- CUTLASS: f3fde58372d33e9a5650ba7b80fc48b3b49d40c8
- Kernel contract v3: four kernels, max M=256, FP32 collapse SSQ, int32 HashMoE input IDs

## Mainline integration

- Based directly on the latest `main` (`678caa988`), with seven MoE-front-only commits.
- Uses the mainline `factory/fused_moe` architecture and publishes through `MegaMoeSEExecutor.forward_prepacked`; it does not restore the removed DSV4-local strategy tree.

## Validation

- Mainline adapter unit tests: 22/22 pass in the SM103 RTP image (PyTorch capability (10, 3)), including learned/hash numerical equivalence, HC writeback, fallback, graph-plan lifecycle, and prepacked empty-rank coverage.
- Exact RTP native build and CUDA 13 extension wheel import together successfully.
- Four-SM103-GPU real-extension CUDA Graph smoke passes.
- Pro learned/hash at M=128: 21.6805/21.3285 us, exactly four kernels.
- Flash learned/hash at M=128: 19.632/19.616 us, exactly four kernels.
- Graph input mutation passes for all four cases.
- Four-rank Decode construction loads the contract-v3 geometry with strategy=mega_moe_se on every rank.
- Paired four-GPU Prefill plus four-GPU Decode validation completes with CSA/HCA off and MoE front on; all Decode ranks capture and replay batch-1 CUDA Graphs successfully, and three routed requests generate 64/64/200 output tokens with pd_sep=true.

The four-GPU evidence above was collected before replaying the seven commits onto `main`. The mainline adapter unit suite has been rerun on SM103; PR CI and a targeted end-to-end SM103 smoke must still revalidate the full engine path. The P/D run uses scratch weights and validates ABI/execution integration only. Real-weight semantic validation, controlled Decode performance A/B, and Decode Perfetto remain follow-up validation.

### CODE REVIEW

**Comment by zsnoob** [comment]
Addressed the current MoE-front review findings in `8f56ace1e`:

- added learned/hash numerical equivalence and final HC writeback coverage;
- made `required` runtime behavior explicit and auto fallback observable;
- aligned `supports()` with the non-null input-ID contract;
- restored pre-FFN debug capture ordering;
- added graph-plan cleanup/all-or-none attach rollback;
- removed dead state/helpers and documented the learned M<=9 logits specialization;
- covered the real `MegaMoeSEExecutor.forward_prepacked` entry for non-empty and empty ranks.

Validation: 22/22 adapter tests pass in the SM103 RTP image; compute capability was verified through `torch.cuda.get_device_capability()` as `(10, 3)`. The PR remains decode-only and preserves the four-kernel front path.

**Comment by zsnoob** [comment]
Addressed the latest AI review findings in `f9a1ffa06`:

- made `required` apply to startup attachment only, while runtime inputs outside the M<=256/tensor contract preserve the ordinary MoE fallback promised by the PR;
- made repeated attachment release the previous adapter/workspace/plans and removed the stale runtime-required state;
- split learned/hash input-ID validation so learned routing does not reject unused non-contiguous IDs, while HashMoE requires contiguous int32 IDs on the residual CUDA device;
- moved the front-path `ffn_in` observer before the routed-expert launch without changing the normal four-kernel path;
- protected all-or-none rollback so cleanup failures no longer mask the original attach error;
- added coverage for repeated attach, M=257 fallback, hash device/dtype constraints, learned non-contiguous IDs, and front debug ordering.

Validation: `git diff --check` and Python compilation pass; 24/24 targeted adapter tests pass in the current-source SM103 harness, with capability verified via `torch.cuda.get_device_capability()` as `(10, 3)`. CUDA extension code and the K1-K4 kernel sequence are unchanged.

File: rtp_llm/models_py/modules/factory/fused_moe/utils/mega_moe/mega_front.py [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+                        norm_eps=float(self.ffn_norm.variance_epsilon),
+                        hc_eps=float(self.ffn_hc.hc_eps),
+                        route_scale=float(self.gate.route_scale),
+                        use_pdl=True,
+                    )
+                else:
+                    assert self.correction_bias is not None
+                    plan.run_learned_out(
+                        self.hc_base,
+                        self.hc_scale,
```
[the remark this answers is not in the document]
Comment (zsnoob) [reply rtp_llm/models_py/modules/factory/fused_moe/utils/mega_moe/mega_front.py:439]: 已核对 cuda-extension 的实际所有权，这里不需要加同步：RTP 始终向 plan 传入 adapter 持有的 workspace/sqr_sum/router_partials，因此 mHC owns_scratch=false、router owns_partials=false；hidden、权重和输出也分别由 caller/model/adapter 持有。close() 只销毁 host context/descriptor wrapper，cudaLaunchKernelExC 在 enqueue 时已复制 kernel 参数，不会释放 kernel 仍引用的 device allocation。374606a42 增加了该所有权说明；增加 synchronize/event 会反而串行化 eager hot path。

File: rtp_llm/models_py/modules/factory/fused_moe/utils/mega_moe/mega_front.py [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+                or input_ids.dtype != torch.int32
+                or not input_ids.is_contiguous()
+            ):
+                raise RuntimeError(
+                    "DSV4 hash MoE front requires contiguous int32 input_ids on "
+                    "the residual CUDA device"
+                )
+            input_ids_flat = input_ids.view(-1)
+
+        if tokens == 0:
```
[the remark this answers is not in the document]
Comment (zsnoob) [reply rtp_llm/models_py/modules/factory/fused_moe/utils/mega_moe/mega_front.py:386]: 已确认生产输入契约是 CUDA int32：cuda_graph_runner.cc:1429-1433 明确校验 input_ids 为 CUDA/int32/contiguous，1780 也用 options_cuda_int32_ 分配 graph input。adapter 内做 .to(int32) 会在每层分配并破坏 graph capture，因此保留零拷贝 int32 契约。374606a42 增加了精确 fallback reason（dtype/device/contiguous/capacity），避免静默降级，并补了测试。

File: rtp_llm/models_py/modules/factory/fused_moe/utils/mega_moe/mega_front.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
...
+            )
+    if build_info.get("kernel_count") != 4:
+        raise RuntimeError(
+            "DSV4 MoE-front build must publish four kernels, got "
+            f"{build_info.get('kernel_count')!r}"
+        )
+
+    source_commit = str(build_info.get("source_commit", ""))
+    source_sha256 = str(build_info.get("source_sha256", ""))
+    if not re.fullmatch(r"[0-9a-f]{8,40}", source_commit):
```
[the remark this answers is not in the document]
Comment (zsnoob) [reply rtp_llm/models_py/modules/factory/fused_moe/utils/mega_moe/mega_front.py:117]: 8-40 位是有意的：cuda-extension/setup.py 的标准 wheel 构建默认使用 git rev-parse --short=8 HEAD；完整实现身份另由 64 位 source_sha256 校验，DeepGEMM/CUTLASS 仍要求 40 位 SHA。374606a42 已补注释和 8 位 source_commit 的契约测试，不能单方面收紧为 40 位，否则会拒绝标准 wheel。

**Comment by zsnoob** [comment]
已推送 374606a42，处理本轮 AI review：补 hidden%128 校验；auto strategy 与 all-or-none 回滚日志；fallback 精确原因；去掉未使用解包；按 device/geometry 缓存扩展加载和 ABI 验证，并在首个 auto attach 失败后停止重复探测。HashMoE 保持生产 CUDA int32 零拷贝契约；source_commit 8-40 位保持与 cuda-extension 标准构建一致。plan-close UAF 经 cuda-extension 所有权代码核对为误报，见对应 inline 回复。sm103-2 上由 PyTorch 确认 capability=(10,3)，当前源码定向测试 26/26 通过；本地 py_compile 与 git diff --check 通过。

**Review by netaddi — APPROVED** [review]

**Review by netaddi — APPROVED** [review]

**Comment by wht21** [comment]
internal source has been updated, please review the changes!

**Comment by wht21** [comment]
internal source has been updated, please review the changes!
